File: backend/app/schemas/physical_datasets.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ColumnInfo(BaseModel):
    name: str
    sample_values: list[str]
    dtype: str


class PreviewResponse(BaseModel):
    columns: list[ColumnInfo]
    preview_rows: list[dict[str, Any]]
    total_rows: int
    suggested_mapping: dict[str, str]
    preview_token: str
    # Первые строки файла «как есть» — для пикера строки заголовков.
    raw_rows: list[list[str]]
    # Применённая строка заголовков (0-based; при первом preview — подсказанная).
    header_row: int


class PreviewReparseRequest(BaseModel):
    preview_token: str = Field(pattern=r"^[0-9a-f]{32}$")
    header_row: int = Field(ge=0)


class PhysicalDatasetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    preview_token: str
    column_mapping: dict[str, Any]
    header_row: int = Field(default=0, ge=0)
    save_as_template: bool = False


class UploadTaskResponse(BaseModel):
    id: int
    status: str
    task_id: str | None


class PhysicalDatasetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    name: str
    file_name: str
    file_size_bytes: int
    status: str
    total_events: int
    total_cases: int
    unique_activities: int
    period_start: datetime | None
    period_end: datetime | None
    health_status: str
    health_report: dict[str, Any]
    column_mapping: dict[str, Any]
    header_row: int
    uploaded_at: datetime
    error_message: str | None


class PhysicalDatasetList(BaseModel):
    items: list[PhysicalDatasetResponse]
    total: int


class TaskStatusResponse(BaseModel):
    task_id: str
    status: str
    result: Any | None = None
    error: str | None = None


class HealthCheckItem(BaseModel):
    name: str
    severity: str
    message: str
    value: Any


class HealthReportResponse(BaseModel):
    status: str
    checks: list[HealthCheckItem]

File: frontend/src/api/physicalDatasets.ts
import { apiClient } from './client';

export interface ColumnInfo {
  name: string;
  sample_values: string[];
  dtype: string;
}

export interface PreviewResponse {
  columns: ColumnInfo[];
  preview_rows: Record<string, unknown>[];
  total_rows: number;
  suggested_mapping: Record<string, string>;
  preview_token: string;
}

export interface PhysicalDataset {
  id: number;
  project_id: number;
  name: string;
  file_name: string;
  file_size_bytes: number;
  status: string;
  total_events: number;
  total_cases: number;
  unique_activities: number;
  period_start: string | null;
  period_end: string | null;
  health_status: string;
  health_report: Record<string, unknown>;
  column_mapping: Record<string, string>;
  uploaded_at: string;
  error_message: string | null;
}

export interface HealthCheckItem {
  name: string;
  severity: string;
  message: string;
  value: unknown;
}

export interface HealthReport {
  status: string;
  checks: HealthCheckItem[];
}

export interface UploadTaskResponse {
  id: number;
  status: string;
  task_id: string | null;
}

export interface CreateDatasetPayload {
  name: string;
  preview_token: string;
  column_mapping: Record<string, string>;
  save_as_template: boolean;
}

export async function previewDataset(
  projectId: number,
  file: File
): Promise<PreviewResponse> {
  const formData = new FormData();
  formData.append('file', file);
  const { data } = await apiClient.post<PreviewResponse>(
    `/projects/${projectId}/physical-datasets/preview`,
    formData
  );
  return data;
}

export async function createDataset(
  projectId: number,
  payload: CreateDatasetPayload
): Promise<UploadTaskResponse> {
  const { data } = await apiClient.post<UploadTaskResponse>(
    `/projects/${projectId}/physical-datasets`,
    payload
  );
  return data;
}

export async function listDatasets(
  projectId: number
): Promise<{ items: PhysicalDataset[]; total: number }> {
  const { data } = await apiClient.get<{ items: PhysicalDataset[]; total: number }>(
    `/projects/${projectId}/physical-datasets`
  );
  return data;
}

export async function getDataset(
  projectId: number,
  datasetId: number
): Promise<PhysicalDataset> {
  const { data } = await apiClient.get<PhysicalDataset>(
    `/projects/${projectId}/physical-datasets/${datasetId}`
  );
  return data;
}

export async function getDatasetHealth(
  projectId: number,
  datasetId: number
): Promise<HealthReport> {
  const { data } = await apiClient.get<HealthReport>(
    `/projects/${projectId}/physical-datasets/${datasetId}/health`
  );
  return data;
}

export async function deleteDataset(
  projectId: number,
  datasetId: number
): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/physical-datasets/${datasetId}`);
}
